### PR TITLE
Use yaml instead of get data patterns when possible

### PR DESCRIPTION
**Pull request by lhoestq**
This would make the data files resolution faster: no need to list all the data files to infer the dataset builder to use.

fix https://github.com/huggingface/datasets/issues/6140

### CODE REVIEW

File: src/datasets/load.py [hunk 1]
```diff
@@ -851,7 +851,12 @@ def get_module(self) -> DatasetModule:
         # even if metadata_configs_dict is not None (which means that we will resolve files for each config later)
         # we cannot skip resolving all files because we need to infer module name by files extensions
         base_path = Path(self.path, self.data_dir or "").expanduser().resolve().as_posix()
-        patterns = sanitize_patterns(self.data_files) if self.data_files is not None else get_data_patterns(base_path)
+        if self.data_files is not None:
+            patterns = sanitize_patterns(self.data_files)
+        if metadata_configs and "data_files" in next(iter(metadata_configs.values())):
+            patterns = sanitize_patterns(next(iter(metadata_configs.values()))["data_files"])
```
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/load.py:857]: This is not supported yet, and would require a refactor.

In a subsequent PR we can raise an error if it happens. Currently it raises an error if it fails to load data files if it uses the wrong dataset builder but this is confusing

**Comment by mariosasko** [comment]
This should also fix https://github.com/huggingface/datasets/issues/6140, so please link it with this PR before merging.

**Comment by lhoestq** [comment]
Done !

**Review by mariosasko — APPROVED** [review]
Oops, forgot to approve 🙂!

PS: The data files resolution (and builder creation) is hard to follow when `metadata_configs` are present. It would be great if we could refactor this eventually to make it easier to maintain this part of the codebase.
